cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextProvider.java

```diff
@@ -16,57 +16,203 @@
+                          kafkaClientService, metricsCollectionService, streamCoordinatorClient);
+      Runtime.getRuntime().addShutdownHook(new Thread() {
+        @Override
+        public void run() {
+          // The logger may already been shutdown. Use System.out/err instead
+          System.out.println("Shutting SparkClassLoader services");
+          Future<List<ListenableFuture<Service.State>>> future = Services.chainStop(logAppenderService,
+                                                                                    streamCoordinatorClient,
+                                                                                    metricsCollectionService,
+                                                                                    kafkaClientService,
+                                                                                    zkClientService);
+          try {
+            List<ListenableFuture<Service.State>> futures = future.get(5, TimeUnit.SECONDS);
+            System.out.println("SparkClassLoader services shutdown completed: " + futures);
+          } catch (Exception e) {
+            System.err.println("Exception when shutting down services");
+            e.printStackTrace(System.err);
+          }
+        }
+      });
+
+      // Create the context object
+      sparkContext = new ExecutionSparkContext(
+        contextConfig.getSpecification(), contextConfig.getProgramId(), contextConfig.getRunId(),
+        classLoader, contextConfig.getLogicalStartTime(), contextConfig.getArguments(),
+        contextConfig.getTransaction(), injector.getInstance(DatasetFramework.class),
+        injector.getInstance(DiscoveryServiceClient.class), metricsCollectionService, hConf,
+        injector.getInstance(StreamAdmin.class)
+      );
+      return sparkContext;
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
     }
-    return context;
   }
 
-  private synchronized AbstractSparkContextBuilder getBuilder(CConfiguration conf) {
-    if (contextBuilder == null) {
-      String mrFramework = hConf.get(MRConfig.FRAMEWORK_NAME, MRConfig.LOCAL_FRAMEWORK_NAME);
-      if ("local".equals(mrFramework)) {
-        contextBuilder = new InMemorySparkContextBuilder(conf);
-      } else {
-        throw new RuntimeException("Spark does not run in distributed mode right now");
+  private static CConfiguration createCConf() throws MalformedURLException {
+    CConfiguration cConf = CConfiguration.create();
+    cConf.clear();
+    cConf.addResource(new File("cConf.xml").toURI().toURL());
+    return cConf;
+  }
+
+  private static Configuration createHConf() throws MalformedURLException {
+    Configuration hConf = new Configuration();
+    hConf.clear();
+    hConf.addResource(new File("hConf.xml").toURI().toURL());
+    return hConf;
+  }
+
+  private static Injector createInjector(CConfiguration cConf, Configuration hConf) {
+    return Guice.createInjector(
+      new ConfigModule(cConf, hConf),
```

Comment:
yes. That's what the check in https://github.com/caskdata/cdap/pull/3094/files#diff-72bffd56ae78926e49675a54f167eab9R99 is for.